### PR TITLE
[Python] Support add column via Expression.

### LINKED ISSUE #321: Append column with an arrow.compute.Expression as value
Allow `dataset.append_column(new_field, expression)`  to mimic SQL syntax, `UPDATE ... SET new_col = expression`.

### PR DESCRIPTION
**Pull request by eddyxu**
Expose `add_column(field, expression)` API via Python.

Closes #321

### CODE REVIEW

File: python/lance/tests/test_schema_evolution.py [hunk 1]
```diff
@@ -30,3 +32,34 @@ def test_write_versioned_dataset(tmp_path: Path):
...
+
+
+def test_add_column_with_compute(tmp_path: Path):
+    table = pa.Table.from_pylist([{"a": i} for i in range(10)])
+
+    base_dir = tmp_path / "test"
+    lance.write_dataset(table, base_dir)
+    dataset = lance.dataset(base_dir)
+    new_dataset = dataset.append_column(pa.field("b", pa.int64()),
+                                        pc.Expression._call("power", [pc.field("a"), pc.scalar(2)]))
```
Comment (changhiskhan) [inline python/lance/tests/test_schema_evolution.py:59]: Why not call pc.power directly? https://arrow.apache.org/docs/python/generated/pyarrow.compute.power.html

File: python/lance/tests/test_schema_evolution.py [hunk 2]
```diff
@@ -30,3 +32,34 @@ def test_write_versioned_dataset(tmp_path: Path):
...
+    new_dataset = dataset.append_column(pa.field("b", pa.float64()), pc.scalar(0.5))
+
+    assert new_dataset.version["version"] == 2
+    actual_df = new_dataset.to_table().to_pandas()
+    expected_df = table.to_pandas()
+    expected_df["b"] = 0.5
+    pd.testing.assert_frame_equal(actual_df, expected_df)
+
+
+def test_add_column_with_compute(tmp_path: Path):
```
Comment (changhiskhan) [inline python/lance/tests/test_schema_evolution.py:52]: There's not test yet for array (pre computed whole column) or function?

File: python/lance/_lib.pyx [hunk 3]
```diff
@@ -273,28 +276,43 @@ cdef class FileSystemDataset(Dataset):
...
+    def _append_column_expr(self, field: Field, expression: Expression) -> FileSystemDataset:
+        cdef CExpression c_expression = expression.unwrap()
+        cdef shared_ptr[CField] c_field = pyarrow_unwrap_field(field)
+        cdef shared_ptr[CLanceDataset] dataset = GetResultValue(self.lance_dataset.AddColumn(c_field, c_expression))
+        return FileSystemDataset.wrap(static_pointer_cast[CDataset, CLanceDataset](dataset))
+
+    def append_column(
+            self,
+            field: Field,
+            value: Union[Callable[[pyarrow.Table], pyarrow.Array], Expression],
```
Comment (changhiskhan) [inline python/lance/_lib.pyx:288]: Scalar is not listed as a supported type?

File: python/lance/_lib.pyx [hunk 4]
```diff
@@ -273,28 +276,43 @@ cdef class FileSystemDataset(Dataset):
...
+    ) -> FileSystemDataset:
         """Append a new column.
 
         Parameters
         ----------
         field : pyarrow.Field
             The name and schema of the newly added column.
-        func : Callback[[pyarrow.Table], pyarrow.Array]
+        value : Callback[[pyarrow.Table], pyarrow.Array], pyarrow.compute.Expression
```
Comment (changhiskhan) [inline python/lance/_lib.pyx:296]: Callable is not tested?

File: python/lance/tests/test_schema_evolution.py (hunk 1 above)
Comment (eddyxu) [reply python/lance/tests/test_schema_evolution.py:59]: `pc.power()` is executing the function. Here we'd like to pass an expression to the runtime to execute later.

File: python/lance/tests/test_schema_evolution.py (hunk 2 above)
Comment (eddyxu) [reply python/lance/tests/test_schema_evolution.py:52]: pre-computed array has not been supported yet.

function test is in this file (from the previous PR)

https://github.com/eto-ai/lance/pull/324/files/1dd095b23e1a7a15cb703a6d6b2b308d420e0198#diff-ed974e93ecd35d82f64c13c4a70f71f6ee757911825654a2856b9d06cf6164c9R24

File: python/lance/_lib.pyx (hunk 4 above)
Comment (eddyxu) [reply python/lance/_lib.pyx:296]: As mentioned above, callable / function was merged from previous PR, w/ their tests.

https://github.com/eto-ai/lance/blob/fd73142c2a1b8a165d2fce1e0da2c68372ded7cd/python/lance/tests/test_schema_evolution.py#L22

File: python/lance/_lib.pyx (hunk 3 above)
Comment (eddyxu) [reply python/lance/_lib.pyx:288]: Scalar is not. `Expression.literal()` is.

**Review by changhiskhan — APPROVED** [review]

**Pull request merged**